scripts/linking.py

```diff
@@ -481,30 +481,43 @@ def get_mention_text_and_ids_by_doc(data: List[data_util.MedMentionExample],
 def get_predicted_mention_candidates_and_types(span,
                                                ner_entities,
                                                filtered_batch_candidate_neighbor_ids,
-                                               predicted_mention_types):
-    candidates = {}
-    mention_types = set()
+                                               predicted_mention_types,
+                                               use_soft_matching):
+    """
+    This function returns three lists, candidates, mention types, and mention spans. These lists will be the same length and have
+    length equal to the number of predicted entities that overlap with the input gold entity. When not using soft mentions,
+    this length will be equal to one, as only one predicted entity can exactly match a gold entity.
+    """
+    candidates = []
+    mention_types = []
+    mention_spans = []
 
     if span is not None:
         for j, predicted_entity in enumerate(ner_entities):
-            overlaps = False
-            # gold span within spacy span
-            if span.start_char >= predicted_entity.start_char and span.end_char <= predicted_entity.end_char \
-                and predicted_entity != span:
-                overlaps = True
-            # spacy span within gold span
-            if predicted_entity.start_char >= span.start_char and predicted_entity.end_char <= span.end_char \
-                and predicted_entity != span:
-                overlaps = True
-            # endpoint overlap between gold span and spacy span
-            if predicted_entity.start_char <= span.start_char and predicted_entity.end_char >= span.start_char \
-                or predicted_entity.start_char <= span.end_char and predicted_entity.end_char >= span.end_char:
-                overlaps = True
-            if overlaps:
-                candidates.update(filtered_batch_candidate_neighbor_ids[j])
-                mention_types.update(predicted_mention_types[j]) 
-
-    return candidates, mention_types
+            if use_soft_matching and span == predicted_entity:
+                candidates = filtered_batch_candidate_neighbor_ids[j]
+                mention_types = predicted_mention_types[j]
+                break
+            elif not use_soft_matching:
```

Comment:
Shouldn't this be `elif use_soft_matching`?

Reply:
and the one above needs to be `if span == predicted_entity:`

Reply:
wow yeah, got that backwards